R/normalize.R

```diff
@@ -15,113 +18,128 @@ normalize <- function(batch_id,
 
-# prepare to load objects by loading image table so we only have to load the data once.
-if (sample_single_cell) {
-  sqlite_file <- paste(full_backend_directory, paste0(plate_id, ".sqlite"), sep = "/")
+  # the columns have a prefix of "Image_" in the SQLite db.
+  well_unique_id_columns_db <- stringr::str_c("Image_", well_unique_id_columns)
 
-  # TODO: if file doesn't exist at path then copy it from S3 to a tmpdir
-  if (!file.exists(sqlite_file)) {
-    stop(paste0(sqlite_file, " does not exist"))
-  }
+  # compartment tag converts nuclei to ^Nuclei_
+  compartment_tag <-
+    function(compartment) paste0("^", stringr::str_to_title(compartment), "_")
 
-  ## cytominer scripts version
-  # db <- dplyr::src_sqlite(path = db_path)
+  # TODO: parse file name as input argument of cytotools_normalize
+  profiles <-
+    readr::read_csv(paste(full_backend_directory,
+                          paste0(plate_id, "_augmented.csv"), sep = "/"))
 
-  db <- DBI::dbConnect(RSQLite::SQLite(), sqlite_file)
+  # prepare to load objects by loading image table so we only have to load the
+  # data once.
 
-  #https://github.com/tidyverse/dplyr/issues/3093
-  RSQLite::initExtension(db)
+  if (sample_single_cell) {
+    sqlite_file <- paste(full_backend_directory,
+                         paste0(plate_id, ".sqlite"), sep = "/")
 
+    # TODO: if file doesn't exist at path then copy it from S3 to a tmpdir
+    if (!file.exists(sqlite_file)) {
+      stop(paste0(sqlite_file, " does not exist"))
+    }
 
-  # get metadata and copy to db
-  metadata <-
-    profiles %>%
-    dplyr::select(dplyr::matches("Metadata_")) %>%
-    dplyr::distinct()
+    db <- DBI::dbConnect(RSQLite::SQLite(), sqlite_file)
 
-  metadata <- copy_to(db, metadata)
+    #https://github.com/tidyverse/dplyr/issues/3093
+    RSQLite::initExtension(db)
 
-  image <- dplyr::tbl(src = db, "image") %>%
-    dplyr::select(TableNumber, ImageNumber, Image_Metadata_Plate, Image_Metadata_Well) %>%
-    dplyr::rename(Metadata_Plate = Image_Metadata_Plate) %>%
-    dplyr::rename(Metadata_Well = Image_Metadata_Well) %>%
-    dplyr::inner_join(metadata, by = c("Metadata_Plate", "Metadata_Well"))
-}
+    # get metadata and copy to db
+    metadata <-
+      profiles %>%
+      dplyr::select(dplyr::matches("Metadata_")) %>%
+      dplyr::distinct()
 
+    metadata <- dplyr::copy_to(db, metadata)
 
-# compartment tag converts nuclei to ^Nuclei_
-compartment_tag <- function(compartment) {
-  stringr::str_c(
-    "^",
-    stringr::str_sub(compartment, 1, 1) %>%
-      stringr::str_to_upper(),
-    stringr::str_sub(compartment, 2),
-    "_"
-  )
-}
+    # TODO: change the renames to use tidyeval form
+    image <- dplyr::tbl(src = db, "image") %>%
+      dplyr::select(c(image_object_join_columns, well_unique_id_columns_db)) %>%
+      dplyr::rename(Metadata_Plate = Image_Metadata_Plate) %>%
+      dplyr::rename(Metadata_Well = Image_Metadata_Well) %>%
```

Comment:
`dplyr::rename("Metadata_Plate" = "Image_Metadata_Plate")` is right! I didn't know that works now.

Regarding the purpose of the renaming itself: that's to strip the `Image_` prefix. Its a hack sadly, and worth revisiting. 

We don't want that prefix because it makes the names too long. We can potentially remove the prefix upstream (in `cytominer-database`) but that could get a little messy. So we need some standard way of pruning that prefix, perhaps by writing a helper functioning and calling it each time? @cells2numbers

Reply:
I like the idea of the helper function! It makes it easy to remove the prefix before  ```aggregate.R``` and we would get consistent data (now we would have different column names in aggregated and normalized data).

One comment: we aggregate the data on per-well basis and than use the aggregated data to compute the mean / median value for the normalization. Should we provide the option to calculate the mean / median values for normalization using single cell data? I am not sure how to implement it but it could be important when only a limited number of controls is available. 
@shntnu

Reply:
I checked that in `cytominer_scripts`, `aggregate`, `annotate` ,`collate`, `normalize`, `stats` all refer to `Image_Metadata_` at some point. 

- `aggregate`, `normalize`, `stats` use it when accessing the `image` table in SQLite backend
- `annotate` uses it to "fix" the column name that `aggregate` produces

Given that `collate` is closest to the SQLite file in terms of its position in the pipeline, the "fix" should be done here. `collate` ingests using `cytominer-database`, then aggregates using `aggregate` , then moves the data to object store. 
(`collate` exists to do all these things at one shot so that the SQLite file needs to be accessed only once. `normalize` and `stats` can use it too, but that's rare)

For now, let's create a script that does the following: it strips the prefix `Image_` for all columns starting with `Image_Metadata`. Then everything downstream of that can assume that the column name does not need to be update.
@cells2numbers

Reply:
> Should we provide the option to calculate the mean / median values for normalization using single cell data? 

You can already do that using `sample_single_cell` in `normalize` @cells2numbers